src/pymor/algorithms/neural_network.py

```diff
@@ -0,0 +1,296 @@
+# This file is part of the pyMOR project (http://www.pymor.org).
+# Copyright 2013-2020 pyMOR developers and contributors. All rights reserved.
+# License: BSD 2-Clause License (http://opensource.org/licenses/BSD-2-Clause)
+
+
+from pymor.core.config import config
+
+
+if config.HAVE_TORCH:
+    import torch
+    import torch.nn as nn
+    import torch.optim as optim
+    import torch.utils as utils
+
+    from pymor.core.base import BasicObject
+    from pymor.core.logger import getLogger
+
+
+    def train_ann(training_data, validation_data, layers,
+            activation_function=torch.tanh, optimizer=optim.LBFGS,
+            epochs=1000, batch_size=20, learning_rate=1., seed=None):
+
+        logger = getLogger('pymor.algorithms.neural_network.train_ann')
+
+        for data in training_data, validation_data:
+            assert isinstance(data, list)
+            assert all(isinstance(datum, tuple) and len(datum) == 2 for datum in data)
+            assert all(isinstance(datum[0], torch.DoubleTensor) for datum in data)
+            assert all(isinstance(datum[1], torch.DoubleTensor) for datum in data)
+
+        # set a seed for the PyTorch initialization of weights and biases and further PyTorch methods
+        if seed:
+            torch.manual_seed(seed)
+
+        # LBFGS-optimizer does not support mini-batching, so the batch size needs to be adjusted
+        if optimizer == optim.LBFGS:
+            batch_size = max(len(training_data), len(validation_data))
+
+        logger.info('Initializing neural network ...')
+
+        # initialize the neural network
+        neural_network = FullyConnectedNN(layers,
+                activation_function=activation_function).double()
```

Comment:
True. Perhaps we make `neural_network=None` optional and create it if `layers` and `activation_function` is given?

Reply:
Sounds good. So as default, we create `neural_network` as `FullyConnectedNN`. Later, one could also think about a `ConvolutionalNN` that is passed to the `train_ann` algorithm directly.

Alternatively, one could choose `NeuralNetworkType=FullyConnectedNN` as parameter for `train_ann` and instantiate the neural network in the training algorithm. Then, I would replace `layers` and `activation_function` with a dictionary `neural_network_parameters` and call something like
```
neural_network = NeuralNetworkType(*neural_network_parameters).double()
```
This would perhaps be the most flexible way if we do not want to instantiate the neural network beforehand. For convolutional neural networks we would call `train_ann(..., NeuralNetworkType=ConvolutionalNN, neural_network_parameters={'layers': [...], 'activation_function': ..., some other parameters specific for convolutional neural networks})`.

Reply:
@ftalbrecht What do you think about this?